src/fraction.ts

```diff
@@ -0,0 +1,88 @@
+import BN from "bn.js";
+
+export class Fraction {
+  private numerator: BN;
+  private denominator: BN;
+
+  constructor(numerator: BN | number, denominator: BN | number) {
+    if (typeof numerator == "number") {
+      this.numerator = new BN(numerator);
+    } else {
+      this.numerator = numerator;
+    }
+
+    if (typeof denominator == "number") {
+      this.denominator = new BN(denominator);
+    } else {
+      this.denominator = denominator;
+    }
+
+    if (this.denominator.isZero()) {
+      throw Error("Denominator cannot be zero");
+    }
+  }
+
+  reduce() {
+    const greatest_common_denominator = gcd(this.numerator, this.denominator);
+    this.numerator = this.numerator.div(greatest_common_denominator);
+    this.denominator = this.denominator.div(greatest_common_denominator);
+  }
+
+  inverted() {
+    return new Fraction(this.denominator, this.numerator);
+  }
+
+  negated() {
+    return new Fraction(this.numerator.neg(), this.denominator);
+  }
+
+  mul(other: Fraction) {
+    const result = new Fraction(
+      this.numerator.mul(other.numerator),
+      this.denominator.mul(other.denominator)
+    );
+    result.reduce();
+    return result;
+  }
+
+  div(other: Fraction) {
+    return this.mul(other.inverted());
+  }
+
+  sub(other: Fraction) {
+    return this.add(other.negated());
+  }
+
+  add(other: Fraction) {
+    const result = new Fraction(
+      this.numerator
+        .mul(other.denominator)
+        .add(other.numerator.mul(this.denominator)),
```

Comment:
uber-tiny-mini-nit:
```suggestion
        .iadd(other.numerator.mul(this.denominator)),
```
Modifies the result of `mul` in-place.